proxy/src/transport/tls/config.rs

```diff
@@ -367,3 +342,886 @@ fn set_common_settings(versions: &mut Vec<rustls::ProtocolVersion>) {
+                            EventMask::DELETE & EventMask::DELETE_SELF & EventMask::CREATE
+                        ) {
+                            self.add_paths()?;
+                        }
+                        return Ok(Async::Ready(Some(())));
+                    },
+                    None => {
+                        debug!("watch stream ending");
+                        return Ok(Async::Ready(None));
+                    },
+                }
+            }
+        }
+    }
+
+    impl Stream for FallbackStream {
+        type Item = ();
+        type Error = ();
+        fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+            self.watch.poll().or_else(|e| {
+                warn!("watch error: {:?}, polling the fs until next change", e);
+                self.polls.poll()
+            })
+        }
+    }
+
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use tempdir::TempDir;
+    use tokio::{
+        runtime::current_thread::Runtime,
+        timer,
+    };
+
+    use std::{
+        path::Path,
+        io::{self, Write},
+        fs::{self, File},
+        thread,
+    };
+
+    use futures::{Sink, Stream};
+    use futures_watch::Watch;
+
+    // These functions are a workaround for Windows having separate API calls
+    // for symlinking files and directories. You're welcome, Brian ;)
+    fn symlink_file<P: AsRef<Path>, Q: AsRef<Path>>(src: P, dst: Q) -> io::Result<()> {
+        #[cfg(target_os = "windows")] {
+            ::std::os::windows::fs::symlink_file(src, dst)?;
+        }
+        #[cfg(not(target_os = "windows"))] {
+            ::std::os::unix::fs::symlink(src, dst)?;
+        }
+        Ok(())
+    }
+
+    fn symlink_dir<P: AsRef<Path>, Q: AsRef<Path>>(src: P, dst: Q) -> io::Result<()> {
+        #[cfg(target_os = "windows")] {
+            ::std::os::windows::fs::symlink_dir(src, dst)?;
+        }
+        #[cfg(not(target_os = "windows"))] {
+            ::std::os::unix::fs::symlink(src, dst)?;
+        }
+        Ok(())
+    }
+
+    struct Fixture {
+        cfg: CommonSettings,
+        dir: TempDir,
+        rt: Runtime,
+    }
+
+    const END_ENTITY_CERT: &'static str = "test-test.crt";
+    const PRIVATE_KEY: &'static str = "test-test.p8";
+    const TRUST_ANCHORS: &'static str = "ca.pem";
+
+    /// A trait that allows an executor to execute a future for up to
+    /// a given time limit, and then panics if the future has not
+    /// finished.
+    ///
+    // TODO: This might be useful for tests outside of this module...
+    trait BlockOnFor {
+        /// Runs the provided future for up to `Duration`, blocking the thread
+        /// until the future completes.
+        fn block_on_for<F>(&mut self, duration: Duration, f: F) -> Result<F::Item, F::Error>
+        where
+            F: Future;
+    }
+
+    impl BlockOnFor for Runtime {
+        fn block_on_for<F>(&mut self, duration: Duration, f: F) -> Result<F::Item, F::Error>
```

Comment:
I think `duration` is maybe better named `timeout`?

Why do we need a timeout at all? If we removed the timeout, would any/all of the tests wait forever?

Reply:
> Why do we need a timeout at all? If we removed the timeout, would any/all of the tests wait forever?

Not when the functionality we're testing for works correctly; however, before everything was working correctly, the failure mode was waiting forever. I found that this made debugging failing tests more difficult, as killing the tests when one has been waiting for over a minute prevents any remaining tests from running, and doesn't print any output from the killed test.

Reply:
OK, I think it's fine to leave the timeouts in. We can remove them later if we find they get in the way. It would be good to add a comment to the constructor function that explains basically what you just said here.

Reply:
Okay, I'll add a comment here.

I was considering moving this to another module, as I think we currently have a number of tests whose failure modes are waiting forever rather than panicking, and changing them to use this might make debugging any regressions that break those tests a bit easier.